apollo-federation/src/type_merging.rs

```diff
@@ -0,0 +1,1364 @@
+                    }
+                    Type::NonNullList(inner2) => {
+                        if !allowed_rules.contains(&SubtypingRule::ListUpgrade) {
+                            return Ok(false);
+                        }
+                        self.is_strict_subtype(type1, inner2)
+                    }
+                    Type::Named(n2) => {
+                        if n1 == n2 {
+                            return Ok(false); // Same type, not a strict subtype
+                        }
+                        if !allowed_rules.contains(&SubtypingRule::Direct) {
+                            return Ok(false);
+                        }
+                        self.is_named_type_subtype(n1, n2)
+                    }
+                    Type::NonNullNamed(n2) => {
+                        if n1 == n2 {
+                            return Ok(false); // Same type, not a strict subtype
+                        }
+                        if !allowed_rules.contains(&SubtypingRule::Direct) {
+                            return Ok(false);
+                        }
+                        self.is_named_type_subtype(n1, n2)
+                    }
+                }
+            }
+            Type::NonNullNamed(n1) => {
+                match type2 {
+                    Type::List(inner2) => {
+                        if !allowed_rules.contains(&SubtypingRule::ListUpgrade) {
+                            return Ok(false);
+                        }
+                        self.is_strict_subtype(type1, inner2)
+                    }
+                    Type::NonNullList(inner2) => {
+                        if !allowed_rules.contains(&SubtypingRule::ListUpgrade) {
+                            return Ok(false);
+                        }
+                        self.is_strict_subtype(type1, inner2)
+                    }
+                    Type::Named(n2) => {
+                        if !allowed_rules.contains(&SubtypingRule::NonNullableDowngrade) {
+                            return Ok(false);
+                        }
+                        if n1 == n2 {
+                            return Ok(true); // NonNull T is subtype of T
+                        }
+                        self.is_named_type_subtype(n1, n2)
+                    }
+                    Type::NonNullNamed(n2) => {
+                        if n1 == n2 {
+                            return Ok(false); // Same type, not a strict subtype
+                        }
+                        if !allowed_rules.contains(&SubtypingRule::Direct) {
+                            return Ok(false);
+                        }
+                        self.is_named_type_subtype(n1, n2)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Check if named_type1 is a subtype of named_type2 (interface/union relationships)
+    fn is_named_type_subtype(
+        &self,
+        subtype: &NamedType,
```

Comment:
Again with the parameter naming, It's unknown whether there is a relationship between these two types, so I think the parameter names should indicate that.